src/textual/widgets/_tabbed_content.py

```diff
@@ -0,0 +1,148 @@
+            id: Optional ID for the TabPane.
+            classes: Optional initial classes for the widget.
+            disabled: Whether the TabPane is disabled or not.
+        """
+        self._title = self.render_str(title)
+        super().__init__(
+            *children, name=name, id=id, classes=classes, disabled=disabled
+        )
+
+
+class TabbedContent(Widget):
+    """A container with associated tabs to toggle content visibility."""
+
+    DEFAULT_CSS = """
+    TabbedContent {
+        height: auto;
+    }
+    TabbedContent > ContentSwitcher {
+        height: auto;
+    }
+    """
+
+    active: reactive[str] = reactive("", init=False)
+    """The ID of the active tab, or empty string if none are active."""
+
+    def __init__(self, *titles: TextType, initial: str = "") -> None:
+        """Initialize a TabbedContent widgets.
+
+        Args:
+            *titles: Positional argument will be used as title.
+            initial: The id of the initial tab, or empty string to select the first tab.
+        """
+        self.titles = [self.render_str(title) for title in titles]
+        self._tab_content: list[Widget] = []
+        self._initial = initial
+        super().__init__()
+
+    def compose(self) -> ComposeResult:
+        """Compose the tabbed content."""
+
+        def set_id(content: TabPane, new_id: str) -> TabPane:
+            """Set an id on the content, if not already present."""
+            if content.id is None:
+                content.id = new_id
+            return content
+
+        # Wrap content in a `TabPane` if required.
+        pane_content = [
+            (
+                set_id(content, f"tab-{index}")
+                if isinstance(content, TabPane)
+                else TabPane(
+                    title or self.render_str(f"Tab {index}"), content, id=f"tab-{index}"
+                )
+            )
+            for index, (title, content) in enumerate(
+                zip_longest(self.titles, self._tab_content), 1
+            )
+        ]
+        # Get a tab for each pane
+        tabs = [
+            ContentTab(content._title, content.id or "") for content in pane_content
+        ]
+        # Yield the tabs
+        yield Tabs(*tabs, active=self._initial)
+        # Yield the content switcher and panes
+        with ContentSwitcher(initial=self._initial):
+            yield from pane_content
+
+    def compose_add_child(self, widget: Widget) -> None:
+        """When using the context manager compose syntax, we want to attach nodes to the switcher."""
+        self._tab_content.append(widget)
+
+    def _on_tabs_tab_activated(self, event: Tabs.TabActivated) -> None:
+        """User clicked a tab."""
+        self.log("on_tabs_tab_activated", event)
+        event.stop()
+        switcher = self.query_one(ContentSwitcher)
+        assert isinstance(event.tab, ContentTab)
+        switcher.current = event.tab.id
+
+    def _on_tabs_tabs_cleared(self, event: Tabs.TabsCleared) -> None:
```

Comment:
We've never really documented the event  handlers. Perhaps because it feels redundant.

If we are going to start, I think we should do it one PR. So I'm inclined to leave it for then.

Reply:
Works for me!